### ISSUE TITLE
Optional<? extends SomeType> not actually optional in generated builder

### ISSUE THREAD

**Issue by rolfw** [immutables/immutables#511]
Given the following immutable and `org.immutables:value:2.3.9`:

```
@Value.Immutable
public interface SomeImmutable {

    Optional<Number> numberOne();

    Optional<? extends Number> numberTwo();
}
```

When I build this type without setting any of the two values like this: `ImmutableSomeImmutable.builder().build();`, I get `IllegalStateException: Cannot build SomeImmutable, some of required attributes are not set [numberTwo]`.

In other words, when an Optional has a parameter with upper bounded wildcard the annotation processor generates a class + builder with a required property, while it should be optional. The other property, `numberOne`, which doesn't have a wildcard is generated correctly and is not required.

Here's the generated code (JavaDoc omitted):
```
@SuppressWarnings({"all"})
@ParametersAreNonnullByDefault
@Generated({"Immutables.generator", "SomeImmutable"})
@Immutable
@CheckReturnValue
public final class ImmutableSomeImmutable implements SomeImmutable {
  private final @Nullable Number numberOne;
  private final Optional<? extends Number> numberTwo;

  private ImmutableSomeImmutable(
      @Nullable Number numberOne,
      Optional<? extends Number> numberTwo) {
    this.numberOne = numberOne;
    this.numberTwo = numberTwo;
  }

  @Override
  public Optional<Number> numberOne() {
    return Optional.ofNullable(numberOne);
  }

  @Override
  public Optional<? extends Number> numberTwo() {
    return numberTwo;
  }

  public final ImmutableSomeImmutable withNumberOne(Number value) {
    @Nullable Number newValue = Preconditions.checkNotNull(value, "numberOne");
    if (this.numberOne == newValue) return this;
    return new ImmutableSomeImmutable(newValue, this.numberTwo);
  }

  public final ImmutableSomeImmutable withNumberOne(Optional<Number> optional) {
    @Nullable Number value = optional.orElse(null);
    if (this.numberOne == value) return this;
    return new ImmutableSomeImmutable(value, this.numberTwo);
  }

  public final ImmutableSomeImmutable withNumberTwo(Optional<? extends Number> value) {
    if (this.numberTwo == value) return this;
    Optional<? extends Number> newValue = Preconditions.checkNotNull(value, "numberTwo");
    return new ImmutableSomeImmutable(this.numberOne, newValue);
  }

  @Override
  public boolean equals(@Nullable Object another) {
    if (this == another) return true;
    return another instanceof ImmutableSomeImmutable
        && equalTo((ImmutableSomeImmutable) another);
  }

  private boolean equalTo(ImmutableSomeImmutable another) {
    return Objects.equal(numberOne, another.numberOne)
        && numberTwo.equals(another.numberTwo);
  }

  @Override
  public int hashCode() {
    int h = 31;
    h = h * 17 + Objects.hashCode(numberOne);
    h = h * 17 + numberTwo.hashCode();
    return h;
  }

  @Override
  public String toString() {
    return MoreObjects.toStringHelper("SomeImmutable")
        .omitNullValues()
        .add("numberOne", numberOne)
        .add("numberTwo", numberTwo)
        .toString();
  }

  public static ImmutableSomeImmutable copyOf(SomeImmutable instance) {
    if (instance instanceof ImmutableSomeImmutable) {
      return (ImmutableSomeImmutable) instance;
    }
    return ImmutableSomeImmutable.builder()
        .from(instance)
        .build();
  }

  public static ImmutableSomeImmutable.Builder builder() {
    return new ImmutableSomeImmutable.Builder();
  }

  @NotThreadSafe
  public static final class Builder {
    private static final long INIT_BIT_NUMBER_TWO = 0x1L;
    private long initBits = 0x1L;

    private @Nullable Number numberOne;
    private @Nullable Optional<? extends Number> numberTwo;

    private Builder() {
    }

    @CanIgnoreReturnValue 
    public final Builder from(SomeImmutable instance) {
      Preconditions.checkNotNull(instance, "instance");
      Optional<Number> numberOneOptional = instance.numberOne();
      if (numberOneOptional.isPresent()) {
        numberOne(numberOneOptional);
      }
      numberTwo(instance.numberTwo());
      return this;
    }

    @CanIgnoreReturnValue 
    public final Builder numberOne(Number numberOne) {
      this.numberOne = Preconditions.checkNotNull(numberOne, "numberOne");
      return this;
    }

    @CanIgnoreReturnValue 
    public final Builder numberOne(Optional<Number> numberOne) {
      this.numberOne = numberOne.orElse(null);
      return this;
    }

    @CanIgnoreReturnValue 
    public final Builder numberTwo(Optional<? extends Number> numberTwo) {
      this.numberTwo = Preconditions.checkNotNull(numberTwo, "numberTwo");
      initBits &= ~INIT_BIT_NUMBER_TWO;
      return this;
    }

    public ImmutableSomeImmutable build() {
      if (initBits != 0) {
        throw new IllegalStateException(formatRequiredAttributesMessage());
      }
      return new ImmutableSomeImmutable(numberOne, numberTwo);
    }

    private String formatRequiredAttributesMessage() {
      List<String> attributes = Lists.newArrayList();
      if ((initBits & INIT_BIT_NUMBER_TWO) != 0) attributes.add("numberTwo");
      return "Cannot build SomeImmutable, some of required attributes are not set " + attributes;
    }
  }
}

**Comment by REWKyleB** [comment]
Thanks for your email. 

This is an auto-reply to inform you that I am currently away from the office, and will be returning on January 4th. I will certainly review your email at that time (or soon after) and will respond as appropriate.

If you are looking for help from REW, you might consider reviewing 
the following channels:

 Need to order something? 
http://www.realestatewebmasters.com/support/ordering-new-work.php/ 
 Need help with the technology? 
http://www.realestatewebmasters.com/support/getting-started.php/

Thanks, 


Thanks! 
-- 
*Kyle Boyd | R&D Programmer* 
Real Estate Webmasters

**Comment by rolfw** [comment]
By the way, I managed to set up a project that allows me to step through the code of various annotation processors during a build, including Immutables; 

Is there any documentation to help me get my bearings in the source code immutables, so that I know where to set breakpoints for debugging myself? I found a few bugs in the mongodb library too that I'd like to investigate.

**Comment by rolfw** [comment]
Hmm, now that I returned to my IDE I noticed a compiler warning that's probably related to this issue: `Warning:(13, 32) java: Wildcards are not supported as elements or key/values. Make it lose its special treatment`

From my perspective this behavior is unexpected though; Is there any reason why types with wildcards can't have special treatment like making them optional?

**Comment by elucash** [comment]
@rolfw thank you for the report!

> From my perspective this behavior is unexpected though; Is there any reason why types with wildcards can't have special treatment like making them optional?

It was made intentionally and, so far, I guess, it still makes sense. The rationale is that if you have a type with a wildcard generic, you just cannot create instances of such a type. I mean that while you can blindly initialize the fields of type `Optional<? extends Type>`, you cannot properly derive the type to initialize it with at shorthand initialization (`Optional<T> attr`, initialized with `builder.attr(T)`). In this case to generate proper code we have to derive `T` type. But we cannot: `? extends Type` is not a valid type. When I say "cannot" I mean that while the generator can generate such a code (using just `Type`) it would be unsound/stupid from the point of the type checking. There is also one important rule to follow: use `? extends` or `? super ` wildcards only on parameter types, never on the return types. I will not able to provide a literature/references to explain this in the details this time, but trust me, you should never do this (wildcards on the return types) and if any "standard" API uses it (like `javax.lang.*`) is clearly a mistake (they probably was experimenting with collection read-only-ness, but this is still a mistake)

> Is there any documentation to help me get my bearings in the source code immutables, so that I know where to set breakpoints for debugging myself? I found a few bugs in the mongodb library too that I'd like to investigate.

It depends. It is generally difficult to debug the annotation processor as it runs at compile time. I usually simply use debug output in the form of sysout or compile warnings. There are tools to move the compilation/processing to a runtime so it should be easy to use plain java debugger: https://github.com/google/compile-testing

**Comment by rolfw** [comment]
@elucash Thanks for the thorough explanation! 

>There is also one important rule to follow: use ? extends or ? super wildcards only on parameter types, never on the return types. I will not able to provide a literature/references to explain this in the details this time, but trust me, you should never do this (wildcards on the return types) and if any "standard" API uses it (like javax.lang.*) is clearly a mistake (they probably was experimenting with collection read-only-ness, but this is still a mistake)

Good point, I didn't realize the implications of that. I see it's actually an item (# 28) in Effective Java:

> **Do not use wildcard types as return types.** Rather than providing
> additional flexibility for your users, it would force them to use
> wildcard types in client code.   
> Properly used, wildcard types are
> nearly invisible to users of a class. They cause methods to accept the
> parameters they should accept and reject those they should reject. **If
> the user of a class has to think about wildcard types, there is
> probably something wrong with the class’s API.**

The reason I'm returning an optional with a wildcard is that the generated builder doesn't accept an Optional of a sub type. Example:

```
public class Main {

    public static void main(String[] args) {

        final SubNumber subNumber = new SubNumber();

        ImmutableSomeImmutable.builder()
                .numberOne(subNumber)
                .numberOne(Optional.of(subNumber))  // <--- Compilation error here
                .build();
    }

    private static class SubNumber extends Number {

        @Override
        public int intValue() {

            throw new UnsupportedOperationException();
        }

        @Override
        public long longValue() {

            throw new UnsupportedOperationException();
        }

        @Override
        public float floatValue() {

            throw new UnsupportedOperationException();
        }

        @Override
        public double doubleValue() {

            throw new UnsupportedOperationException();
        }
    }
}
```
How should I go about that?

**Comment by rolfw** [comment]
> It depends. It is generally difficult to debug an annotation processor as it runs at compile time. I usually simply use debug output in the form of sysout or compile warnings. There are tools to move the compilation/processing to a runtime so it should be easy to use plain java debugger: https://github.com/google/compile-testing

Thank you! I'll have a look at compile-testing. Where do you generally recommend to set a breakpoint for the value lib, as well as mongodb and gson generation? For instance, I'd like to investigate why a Gson type adapter is being generated even though I don't have an `@Gson.TypeAdapters` annotation present. So I'd like to find the point where a particular class is generated so that I can see if it's a bug or a mistake of mine.

**Comment by rolfw** [comment]
Getting off-topic here :) I'll re-post it in Google Groups

**Comment by elucash** [comment]
I'll start with the topic )
So to immediately fix the example, you can qualify type parameter `.numberOne(Optional.<Number>of(subNumber))` but it will not solve the general case. For the general case I'm considering we might relax `init(Optional<Number>)` to accept `init(Option<? extends Number>)` as we do with collection initializers like `addAll`.

**Comment by Stephan202** [comment]
:+1: That sounds like a solution in line with [PECS](https://stackoverflow.com/questions/2723397/what-is-pecs-producer-extends-consumer-super).

**Comment by rolfw** [comment]
Thanks, that would be great! I'll use the workaround for the time being

**Comment by elucash** [comment]
released in 2.4.0, thank you!
